File: modules/ROOT/examples/Columnar.scala
import org.apache.spark.sql.SparkSession

object Columnar {
  // tag::init[]
  val spark = SparkSession
    .builder()
    .appName("CouchbaseColumnarSample") // your app name
    .master("local[*]") // your local or remote Spark master node
    .config("spark.couchbase.connectionString", "couchbases://your-columnar-endpoint.cloud.couchbase.com")
    .config("spark.couchbase.username", "username")
    .config("spark.couchbase.password", "password")
    .getOrCreate()
  // end::init[]

  // tag::reading-dataframe[]
  val airlines = spark.read
    .format("couchbase.columnar")
    .option(ColumnarOptions.Database, "travel-sample")
    .option(ColumnarOptions.Scope, "inventory")
    .option(ColumnarOptions.Collection, "airline")
    .load
  // end::reading-dataframe[]


  // tag::reading-dataset[]
  case class Airline(id: String, name: String, country: String) // (1)

  val sparkSession = spark
  import sparkSession.implicits._ // (2)

  val airlinesDataset = spark.read
    .format("couchbase.columnar")
    .option(ColumnarOptions.Database, "travel-sample")
    .option(ColumnarOptions.Scope, "inventory")
    .option(ColumnarOptions.Collection, "airline")
    .load
    .as[Airline] // (3)
  // end::reading-dataset[]

  // tag::processing-dataframe[]
  println(airlines.count)

  airlines.foreach(row => {
    val id = row.getAs[String]("id")
    val name = row.getAs[String]("name")
    println(s"Row: id=${id} name=${name}")
  })
  // end::processing-dataframe[]

  // tag::writing-dataframe[]
  airlines.write
    .format("couchbase.columnar")
    .option(ColumnarOptions.Database, "targetDatabase")
    .option(ColumnarOptions.Scope, "targetScope")
    .option(ColumnarOptions.Collection, "targetCollection")
    .save()
  // end::writing-dataframe[]

  // tag::sql[]
  airlines.createOrReplaceTempView("airlinesView")
  val airlinesFromView = spark.sql("SELECT * FROM airlinesView")
  // end::sql[]

}

File: modules/ROOT/examples/Columnar.py
from pyspark.sql import SparkSession

# tag::init[]
spark = (SparkSession.builder
    .appName("Couchbase Spark Connector Columnar Example")
    # Note whether you need the .master(...) and .config("spark.jars"...) lines depends on how you are using Spark.
    # See our PySpark documentation for more details.
    .master("local[*]")
    .config("spark.jars", """C:\\Users\\Graham Pople\\Downloads\\Couchbase-Spark-Connector_2.12-3.5.1\\spark-connector-assembly-3.5.1.jar""")
    .config("spark.couchbase.connectionString", "couchbases://cb.kiofzclm8ogfldf.customsubdomain.nonprod-project-avengers.com")
    .config("spark.couchbase.username", "test")
    .config("spark.couchbase.password", "Password!1")
    .config("spark.ssl.insecure", "true")
    .getOrCreate())
# end::init[]

# tag::reading-dataframe[]
airlines = (spark.read
    .format("couchbase.columnar")
    .option("database", "travel-sample")
    .option("scope", "inventory")
    .option("collection", "airline")
    .load())
# end::reading-dataframe[]


# tag::processing-dataframe[]
airlines.show()
print(airlines.count())

collected = airlines.collect()

for airline in collected:
    print(airline)
    id = airline["id"]
    name = airline["name"]
    print(f"Airline: id={id} name={name}")
# end::processing-dataframe[]

# tag::sql[]
airlines.createOrReplaceTempView("airlinesView")
airlinesFromView = spark.sql("SELECT * FROM airlinesView")
print(airlinesFromView.count())
# end::sql[]
